src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/datasource/OffenderRisksDataSource.kt

```diff
@@ -195,11 +197,104 @@ class CommunityApiOffenderRisksDataSource(
-    throw NotImplementedError("Getting risks for individual offenders from the AP Delius Context API is not currently supported")
+    val caseDetailResponse = apDeliusContextApiClient.getCaseDetail(crn)
+
+    return PersonRisks(
+      roshRisks = getRoshRisksEnvelope(crn),
+      mappa = caseDetailResponse.toMappa(),
+      tier = getRiskTierEnvelope(crn),
+      flags = caseDetailResponse.toRiskFlags(),
+    )
+  }
+
+  private fun getRoshRisksEnvelope(crn: String): RiskWithStatus<RoshRisks> {
+    return when (val roshRisksResponse = apOASysContextApiClient.getRoshRatings(crn)) {
+      is ClientResult.Success -> {
+        val summary = roshRisksResponse.body.rosh
+
+        when (summary.anyRisksAreNull()) {
+          true -> RiskWithStatus(status = RiskStatus.NotFound)
+          false -> RiskWithStatus(
+            value = RoshRisks(
+              overallRisk = summary.determineOverallRiskLevel().text,
+              riskToChildren = summary.riskChildrenCommunity!!.text,
+              riskToPublic = summary.riskPublicCommunity!!.text,
+              riskToKnownAdult = summary.riskKnownAdultCommunity!!.text,
+              riskToStaff = summary.riskStaffCommunity!!.text,
+              lastUpdated = roshRisksResponse.body.dateCompleted?.toLocalDate()
+                ?: roshRisksResponse.body.initiationDate.toLocalDate(),
+            ),
+          )
+        }
+      }
+      is ClientResult.Failure.StatusCode -> when (roshRisksResponse.status) {
+        HttpStatus.NOT_FOUND -> RiskWithStatus(status = RiskStatus.NotFound)
+        else -> RiskWithStatus(status = RiskStatus.Error)
+      }
+      is ClientResult.Failure -> RiskWithStatus(status = RiskStatus.Error)
+    }
+  }
+
+  private fun getRiskTierEnvelope(crn: String): RiskWithStatus<RiskTier> =
+    when (val tierResponse = hmppsTierApiClient.getTier(crn)) {
+      is ClientResult.Success -> RiskWithStatus(
+        value = RiskTier(
+          level = tierResponse.body.tierScore,
+          lastUpdated = tierResponse.body.calculationDate.toLocalDate(),
+        ),
+      )
+      is ClientResult.Failure.StatusCode -> when (tierResponse.status) {
+        HttpStatus.NOT_FOUND -> RiskWithStatus(status = RiskStatus.NotFound)
+        else -> RiskWithStatus(status = RiskStatus.Error)
+      }
+      is ClientResult.Failure -> RiskWithStatus(status = RiskStatus.Error)
+    }
+
+  fun ClientResult<CaseDetail>.toMappa(): RiskWithStatus<Mappa> = when (this) {
+    is ClientResult.Success -> this.body.toMappa()
+    is ClientResult.Failure.StatusCode -> when (this.status) {
+      HttpStatus.NOT_FOUND -> RiskWithStatus(status = RiskStatus.NotFound)
+      else -> RiskWithStatus(status = RiskStatus.Error)
+    }
+    is ClientResult.Failure -> RiskWithStatus(status = RiskStatus.Error)
+  }
+
+  fun CaseDetail.toMappa(): RiskWithStatus<Mappa> = this.mappaDetail?.let {
+    RiskWithStatus(
+      value = Mappa(
+        "CAT ${it.categoryDescription}/LEVEL ${it.levelDescription}",
```

Comment:
👋 The reason CAS3 [recently changed this to add a guard](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1215/files#diff-4a8c3767d589a6283081bfd38ef72be5ebc2dd5ec9d401127ded27adb5ec015cR567) was because [we found bad data that blocked a new referral from being created](https://trello.com/c/mSSzt3eH). When we tried to load in the risk info into a brand new application, we got a 500 when unpacking the data because the required `category` was missing.

As we now are we aren't using `!!` I guess we wouldn't 500 but the value would be stripped back to a pretty useless "CAT /LEVEL"? In those cases I think it's best we throw an error instead and allow the frontend to display the "We couldn't find data for this right now" etc in the risk widget.

Though the names of these values have changed I should think we ought to continue guarding against this and other values being nil?

Reply:
It could well be worth creating a CAS1 and CAS3 referral locally with this change and use the CRNs that have missing data to test how this works.

Reply:
I couldn't do a local manual test with the list of CRNs that you sent me, but I've implemented a check to make sure that `categoryDescription` and `levelDescription` are both not null. If either are, it'll return the `NotFound` variant of the `RiskWithStatus`.